doc/devdocs/locks.rst

```diff
@@ -0,0 +1,110 @@
+.. currentmodule:: Base
+
+****************************************************
+Proper maintenance and care of multi-threading locks
+****************************************************
+
+The following strategies are used to ensure that the code is dead-lock free:
+
+ 1. structure code such that only one lock will need to be acquired at a time
+
+ 2. always acquire shared locks in the same order, as given by the table below
+
+ 3. avoid constructs that expect to need unrestricted recursion
+
+Below are all of the locks that exist in the system and the mechanism in use
+that prevents them from being a potential deadlock.
+
+The following are definitely leaf locks (level 1), and must not try to acquire any other lock:
+
+   * safepoint
+
+       note that this lock is acquired implicitly by ``JL_LOCK`` and ``JL_UNLOCK``.
+       use the ``_NOGC`` variants to avoid that for level 1 locks.
+
+   * shared_map
+   * finalizers
+   * pagealloc
+   * flisp
+
+       flisp itself is already threadsafe, this lock only protects the ``jl_ast_context_list_t`` pool
+
+
+The following is a leaf lock (level 2), and only acquires level 1 locks (safepoint) internally:
+
+   * symbol_table
+
+The following are level 3 locks, which can only acquire level 1 or level 2 locks internally:
+
+   * dump
```

Comment:
Depending on how likely we'll have this running on multiple thread (maybe not if we can directly work on bytecode) we can probably get rid of this one fairly easily by storing the states on stack.